File: it-agents/skills/code-architect/references/code-examples/java/Strategy.java
// Strategy — a family of interchangeable algorithms behind one interface.

import java.util.Arrays;

interface SortStrategy {
    int[] sort(int[] input);
}

class AscendingStrategy implements SortStrategy {
    public int[] sort(int[] input) {
        int[] out = input.clone();
        Arrays.sort(out);
        return out;
    }
}

class DescendingStrategy implements SortStrategy {
    public int[] sort(int[] input) {
        int[] out = input.clone();
        Arrays.sort(out);
        for (int i = 0, j = out.length - 1; i < j; i++, j--) {
            int t = out[i]; out[i] = out[j]; out[j] = t;
        }
        return out;
    }
}

class Sorter {
    private SortStrategy strategy;
    void setStrategy(SortStrategy s) { this.strategy = s; }
    int[] run(int[] data) { return strategy.sort(data); }
}

public class Strategy {
    public static void main(String[] args) {
        int[] data = { 3, 1, 4, 1, 5 };
        Sorter sorter = new Sorter();

        sorter.setStrategy(new AscendingStrategy());
        System.out.println("asc:  " + Arrays.toString(sorter.run(data)));

        sorter.setStrategy(new DescendingStrategy());
        System.out.println("desc: " + Arrays.toString(sorter.run(data)));
    }
}

File: it-agents/skills/code-architect/references/code-examples/python/strategy.py
"""Strategy — a family of interchangeable algorithms behind one interface."""

from __future__ import annotations

from abc import ABC, abstractmethod


class SortStrategy(ABC):
    @abstractmethod
    def sort(self, data: list[int]) -> list[int]:
        ...


class AscendingStrategy(SortStrategy):
    def sort(self, data: list[int]) -> list[int]:
        return sorted(data)


class DescendingStrategy(SortStrategy):
    def sort(self, data: list[int]) -> list[int]:
        return sorted(data, reverse=True)


class Processor:
    """Holds a strategy and defers the algorithm choice to it."""

    def __init__(self, strategy: SortStrategy) -> None:
        self._strategy = strategy

    def set_strategy(self, strategy: SortStrategy) -> None:
        self._strategy = strategy

    def run(self, data: list[int]) -> list[int]:
        return self._strategy.sort(data)


if __name__ == "__main__":
    processor = Processor(AscendingStrategy())
    print("asc: ", processor.run([3, 1, 2]))
    processor.set_strategy(DescendingStrategy())
    print("desc:", processor.run([3, 1, 2]))
